Tad reporting standard (#93)
MinBZK/ai-validation: docs/Projects/TAD/reporting_standard.md

the Hugging Face metadata.

## Specification of the standard

The standard will be written in yaml. An example yaml is given in the next section. The standard contains the
following information.

### `system_card`

A `system_card` contains the following information.

1. `upl` (OPTIONAL, string). If this algorithm is part of a product offered by the Dutch Government,
    it should contain a URI from the [Uniform Product List](https://standaarden.overheid.nl/owms/oquery/UPL-actueel.plain).
2. `owners` (list). There can be multiple owners. For each owner the following fields are present.

    1. `organization` (REQUIRED, string). Name of the organization that owns the model.
    2. `oin` (OPTIONAL, string). If applicable the [Organisatie-identificatienummer (OIN)](https://www.logius.nl/domeinen/toegang/organisatie-identificatienummer/wat-is-het).
    3. `name` (OPTIONAL, string). Name of a contact person within the organisation.
    4. `email` (OPTIONAL, string). Email address of the contact person or organization.
    5. `role` (OPTIONAL, string). Role of the contact person. This field should only be set when the `name` field
    is set.

#### 2. Models

> 1. `models` (OPTIONAL, list[string]). A URI to a yaml file containg a model card. This model card can for example be a model
> card described in the next section or a model card from Hugging Face. There can be multiple model cards, meaning multiple
> models are used.

Reviewer: ```suggestion
1. `models` (OPTIONAL, list[ModelCard]). A list of model cards (as defined below) or `!include`s of a yaml file containing a model card. This model card can for example be a model
card described in the next section or a model card from Hugging Face. There can be multiple model cards, meaning multiple
models are used.
```

    There can be multiple languages.
3. `license`(REQUIRED, string). Any license from the [open source license list](https://opensource.org/license)
[^1]. If the license is NOT present in the license list this field must be set to 'other' and the following
two fields will be REQUIRED.

    1. `license_name` (string). An id for the license.
    2. `license_link` (string). A link to a file of that name inside the repo, or a URL to a remote file containing the license
    contents.

4. `library_name` (OPTIONAL, string). Any library from the [library list](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries.ts).
    If the library is NOT present in the library list, the following fields CAN be set.
    1. `library` (OPTIONAL, string). If a library is used that is not in the above list, you can provide
    the name of the library used here.
    2. `library_link` (OPTIONAL, string). If a libary is used that is not in the above list, you can
    provide a URI to the repo of the library used here.
5. `tags` (OPTIONAL, list[string]). Tags with keywords to describe the project. There can be multiple tags.

> ### 2. `model_index`

Reviewer 2: ```suggestion
#### `model_index`
```
I think this is a typo, should be one level under model (maybe even on the same level as `tags` but i get you want to have special highlight for this and therefore to make it big as it can be source of discussion

# TAD Reporting Standard

> Version 0.1

Reviewer 3: Or with label `alpha` to show this is not an "official" version yet:
```suggestion
*Version 0.1a1*
```

and [ALTAI](https://digital-strategy.ec.europa.eu/en/library/assessment-list-trustworthy-artificial-intelligence-altai-self-assessment)).
This is achieved by defining a new field `assessments`.

This standard does not contain all fields present in the Hugging Face metadata specification. The fields that
are optional in the HuggingFace specification AND are specific to the HugginFace interface are ommited.

Following Hugging Face, this proposed standard will be written in yaml.

Another difference is that we devide our implementation into two parts.

1. `systems_card`, containing information about a group of ML-models which accomplish a specific task.
2. `model_card`, containing information about a specific ML-model.

## Intended usage

The reporting standard defined in this document is intented to be given as a preamble in a README, similar to
the Hugging Face metadata.

## Specification of the standard

The standard will be written in yaml. An example yaml is given in the next section. The standard defines
two 'cards': a `system_card` and a `model_card`.

### `system_card`

> A `system_card` contains the following information.

Author: The model card (not the system card) contains a field algorithm-registers to refer to for example the algoritmeregister. Doesn't this make this information redundant?

    contents.

4. `library_name` (OPTIONAL, string). Any library from the [library list](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries.ts).
    If the library is NOT present in the library list, the following fields CAN be set.
    1. `library` (OPTIONAL, string). If a library is used that is not in the above list, you can provide
    the name of the library used here.
    2. `library_link` (OPTIONAL, string). If a libary is used that is not in the above list, you can
    provide a URI to the repo of the library used here.
5. `tags` (OPTIONAL, list[string]). Tags with keywords to describe the project. There can be multiple tags.

#### `model_index`

There can be multiple models. For each model the following fields are present.

1. `name` (REQUIRED, string). The name of the model.
2. `model` (REQUIRED, string). A URI pointing to a repository containing the model file.
3. `artifacts` (OPTIONAL, list[string]). A list of URI's where each URI refers to a relevant model
artifact, that cannot be captured by any other field, but are relevant to model.

> 4. `parameters` (list). There can be multiple parameters. For each parameter the following fields are present.

Reviewer 3: I would propose to also add `labels` here. F.e. a parameter can come from "a source" and  you would like to have a way to store this meta information. Labels can be a way to do this.

Author: Do you mean a labels field under the parameters field or a labels field on the same level als the parameters fields?

                1. `class` (OPTIONAL, string/int/float/bool). The output class name that the graph corresponds to.
                This field is not always present.
                2. `feature` (REQUIRED, string). The feature the graph corresponds to. This is required, since all
                relevant graphs are dependend on features.
                3. `data` (list)
                    1. `x_value` (REQUIRED, float). The $x$-value of the graph.
                    2. `y_value` (REQUIRED, float). The $y$-value of the graph.

## Example

### System Card

```yaml
upl: {upl_uri}                                          # Optional. Example: https://standaarden.overheid.nl/owms/terms/AangifteVertrekBuitenland
owners:
- organization: {organization_name}                     # Required. Example: BZK
  oin: {oin}                                            # Optional. Example: 00000001003214345000
  name: {owner_name}                                    # Optional. Example: John Doe
  email: {owner_email}                                  # Optional. Example: johndoe@email.com
  role: {owner_role}                                    # Optional. Example: Data Scientist.

# Model Cards
models:

>  - !include {model_card_uri}                            # Optional. Example: model_card.yaml.

Reviewer 3: Curious: is this an supported way of including yaml? Or should we consider using `include:` with a reference to a model card file or model_card_uri?

Author: There is no standard way of include's or import's in yaml that is out-of-the-box supported. There exists libraries to handle things like !include.

# Assessments like IAMA.
assessments:
- name: {assessment_name}                               # Required. Example: IAMA.
  date: {assessment_date}                               # Required. Example: 25-03-2025.
  contents:
    - question: {question_text}                         # Required. Example: "Question 1: ...".
      answer: {answer_text}                             # Required. Example: "Answer: ...".
      remarks: {remarks_text}                           # Optional. Example: "Remarks: ...".
      authors:                                          # Optional. Example: "['John', 'Peter']".
        - name: {author_name}
      timestamp: {timestamp}                            # Optional. Example: 1711630721.
```

> ### Model Card

Reviewer 3: In the category "how far do we want to go for the first itterarion?". We could add a section `system_info` that stores information about the system used. F.e. nr CPUs, GPUs, etc.

Author: Agreed to not include this at this point.

```yaml
algorithm-registers:
  - {id}                                                # Optional. Example: https://algoritmes.overheid.nl/nl/algoritme/45991543.
language:
  - {lang_0}                                            # Optional. Example nl.
license: {licence}                                      # Required. Example: Apache-2.0 or any license SPDX ID from https://opensource.org/license or "other".
license_name: {licence_name}                            # Required. if license = other, specify an id for the licence. Example: 'my-license-1.0'
license_link: {license_link}                            # Required. if license = other, specify "LICENSE" or "LICENSE.md" to link to a file of that name inside the repo, or a URL to a remote file.

> library_name: {library_name}                            # Optional. Example: keras or any library from https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries.ts.

Reviewer 3: Shall we apply same "setup" for license and library? Here it would mean we should rename `library_name` -> `library` where the `library` can contain a value from the mentioned source (similar to what we do with license).

And `library` will be renamed to `library_name`.

Author: This is what I meant yesterday that the naming can become awkward by adhering to the original names of HF. I also would like to have consistent naming, but then we also lose the compatibility with HF for the library field. Let's also discuss this in the meeting you proposed.

Author: We decided to remove the library reference.

      name: {task_name}                                 # Optional. Example: Image Classification.
    dataset:
      - type: {dataset_type}                            # Required. Example: common_voice. Link to a repository containing the dataset
        name: {dataset_name}                            # Required. Example: "Common Voice (French)". A pretty name for the dataset.
        split: {split}                                  # Optional. Example "train".
        features:
         - {feature_name}                               # Optional. Example: "gender".
        revision: {dataset_version}                     # Optional. Example: 5503434ddd753f426f4b38109466949a1217c2bb
    metrics:
    - type: {metric_type}                               # Required. Example: false-positive-rate. Use metric id from https://hf.co/metrics.
      name: {metric_name}                               # Required. Example: "FPR wrt class 0 restricted to feature gender:0 and age:21".
      dtype: {metric_dtype}                             # Required. Example: "float".
      value: {metric_value}                             # Required. Example: 0.75.

>       class: {metric_output_class}                      # Optional. Example: "1". Only relevant for certain metrics such as for example false-positive-rate positive rate.

Reviewer 3: Could give an example how we use this i.c.w. dtype and value? Or could we achieve logging the same information with `value` i.c.w. `dtype`?

      labels:
        - name: {feature_name}                          # Optional. Example: "gender".

>           type: {feature_type}                          # Optional. Example: "string".

Reviewer 3: ```suggestion
          dtype: {feature_type}                          # Optional. Example: "string".
```

to allow for:

1. More finegrained information on performance metrics, by extending the `metrics_field` from the Hugging
Face metadata specification.
2. Capturing additional *measurements* on fairness and bias, which can be partitioned into bar plot like
measurements (such as mean absolute SHAP values) and graph plot like measurements (such as partial dependence). This is achieved
by defining an new field `measurements`.
3. Capturing *assessments* (such as [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)
and [ALTAI](https://digital-strategy.ec.europa.eu/en/library/assessment-list-trustworthy-artificial-intelligence-altai-self-assessment)).
This is achieved by defining a new field `assessments`.

Following Hugging Face, this proposed standard will be written in yaml.

This standard does not contain all fields present in the Hugging Face metadata specification. The fields that
are optional in the HuggingFace specification and are specific to the HugginFace interface are ommited.

Another difference is that we devide our implementation into three seperate parts.

> 1. `systems_card`, containing information about a group of ML-models which accomplish a specific task.

Reviewer: Is it systems_card or system_card?

Author: system_card

2. `model_card`, containing information about a specific ML-model.
3. `assessment_card`, containing information about a regulatory assessment.

> 

Reviewer: ```suggestion
!!! note "Include statements"

    These `model_card`s and  `assessment_card`s  can be included verbatim into a `systems_card`, or referenced with an `!include` statement, allowing for minimal cards to be compact and in a single file, but for extensive cards to be split up for readability and maintainability. 

    Our standard allows for the `!include` to be used anywhere 

```

## Intended usage

The reporting standard defined in this document is intented to be given as a preamble in a README, similar to
the Hugging Face metadata.

## Specification of the standard

The standard will be written in yaml. Example yaml files are given in the next section. The standard defines
three cards: a `system_card`, a `model_card` and an `assessment_card`. A `system_card` contains information
about an algorithmic system. It can have mutiple models and each of these models should have a `model_card`.
Regulatory assessments can be processed in an `assessment_card`. Note that `model_card`'s and
`assessment_card`'s can be included directly into the `system_card` or can be included as seperate yaml
files with help of a yaml-include mechanism. For clarity the latter is preffered and is also used in
the examples in the next section.

### `system_card`

A `system_card` contains the following information.

> 1. `name` (OPTIONAL, string). Name used to describe the system.

Reviewer: ```suggestion
0. `schema_version` (REQUIRED, string). Version of the schema used, for example "0.1a1".
1. `name` (OPTIONAL, string). Name used to describe the system.
```

# TAD Reporting Standard

**Version**: 0.1a1

This document describes the **T**ransparency of **A**lgorithmic **D**ecision making (TAD) Reporting Standard.

For reproducibility, governance, auditing and sharing of algorithmic systems it is essential to have a
reporting standard so that information about an algorithmic system can be shared. This reporting standard
describes how information about the different phases of an algorithm's life cycle can be reported.
It contains, among other things, descriptive information combined with information about the technical
tests and assessments applied.

> ## Disclaimer
> 
> The TAD Reporting Standard is work in progress. This means that the current standard is probably suboptimal
> and will change significantly in future versions.

Reviewer 3: What do you think of making this a warning?

```suggestion
!!! warning "Disclaimer"

    The TAD Reporting Standard is work in progress. This means that the current standard is probably suboptimal
    and will change significantly in future versions.
```

Face metadata specification.
2. Capturing additional *measurements* on fairness and bias, which can be partitioned into bar plot like
measurements (such as mean absolute SHAP values) and graph plot like measurements (such as partial dependence). This is achieved
by defining an new field `measurements`.
3. Capturing *assessments* (such as [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)
and [ALTAI](https://digital-strategy.ec.europa.eu/en/library/assessment-list-trustworthy-artificial-intelligence-altai-self-assessment)).
This is achieved by defining a new field `assessments`.

Following Hugging Face, this proposed standard will be written in yaml.

This standard does not contain all fields present in the Hugging Face metadata specification. The fields that
are optional in the HuggingFace specification and are specific to the HugginFace interface are ommited.

Another difference is that we devide our implementation into three seperate parts.

1. `system_card`, containing information about a group of ML-models which accomplish a specific task.

> 2. `model_card`, containing information about a specific ML-model.

Reviewer 3: Feel free to ignore this suggestion:

```suggestion
2. `model_card`, containing information about a specific data science model.
```

3. `assessment_card`, containing information about a regulatory assessment.

!!! note "Include statements"

>     These `model_card`s and  `assessment_card`s  can be included verbatim into a `system_card`,
>     or referenced with an `!include` statement, allowing for minimal cards to be compact and in a
>     single file, but for extensive cards to be split up for readability and maintainability.
>     Our standard allows for the `!include` to be used anywhere.

Reviewer 3: ```suggestion
    These `model_card`s and  `assessment_card`s  can be included verbatim into a `system_card`,
    or referenced with an `!include` statement, allowing for minimal cards to be compact in a single
    file. Extensive cards can be split up for readability and maintainability. Our standard allows for
    the `!include` to be used anywhere.
```

> ## Intended usage
> 
> The reporting standard defined in this document is intented to be given as a preamble in a README, similar to
> the Hugging Face metadata.
> 

Reviewer 3: I want to propose to remove this section for now. In a new version we can add examples of how it can be used.
```suggestion
```

        results:
         - class: {class_name}                          # Optional. Name of the output class the graph depends on.
           feature: {feature_name}                      # Required. Name of the feature the graph depends on.
           data:
            - x_value: {x_value}                        # Required. The x value of the graph data.
              y_value: {y_value}                        # Required. The y value of the graph data.

```

### Assessment Card

```yaml
name: {assessment_name}                               # Required. Example: IAMA.
date: {assessment_date}                               # Required. Example: 25-03-2025.
contents:
  - question: {question_text}                         # Required. Example: "Question 1: ...".
    answer: {answer_text}                             # Required. Example: "Answer: ...".
    remarks: {remarks_text}                           # Optional. Example: "Remarks: ...".
    authors:                                          # Optional. Example: "['John', 'Peter']".
      - name: {author_name}
    timestamp: {timestamp}                            # Optional. Example: 1711630721.

> ````

Reviewer 3: ```suggestion
```
```

Reviewer 3: @ChristopherSpelt here I tried to suggest the fix to go from 4x grave to 3x grave

## Schema

> JSON schema will be added here.

Reviewer 3: ```suggestion
JSON schema will be added when we publish the first beta version.
```

to allow for:

1. More finegrained information on performance metrics, by extending the `metrics_field` from the Hugging
Face metadata specification.
2. Capturing additional *measurements* on fairness and bias, which can be partitioned into bar plot like
measurements (such as mean absolute SHAP values) and graph plot like measurements (such as partial dependence). This is achieved
by defining an new field `measurements`.
3. Capturing *assessments* (such as IAMA). This is achieved by defining a new field `assessments`.

This standard does not contain all fields present in the Hugging Face metadata specification. The fields that
are optional in the HuggingFace specification AND are specific to the HugginFace interface are ommited.

Following Hugging Face, this proposed standard will be written in yaml.

## Intended usage

The reporting standard defined in this document is intented to be given as a preamble in a model card, similar to
the Hugging Face metadata.

## Specification of the standard

The standard will be written in yaml. An example yaml is given in the next section. The standard contains the
following information.

> 

Reviewer: ```suggestion

0. `algorithm-registers`
    1. `id` (OPTIONAL, URI). If this algorithm is registered in an Algorithm register, this field should contain its ID in that register
```

> 1. `language` (OPTIONAL, list[string]). If relevant, the languages the model support. There can be multiple languages.

Reviewer: ```suggestion
1. `language` (OPTIONAL, list[string]). If relevant, the natural languages the model support. There can be multiple languages.
```

Reviewer 4: ```suggestion
1. `language` (OPTIONAL, list[string]). If relevant, the languages the model support in ISO 639. There can be multiple languages.
```

the license is NOT present in the license list, the following two fields will be REQUIRED.

    1. `license_name` (string). An id for the license.
    2. `license_link` (string). A link to a file of that name inside the repo, or a URL to a remote file containing the license contents.

3. `library` (OPTIONAL, string). Any library from the [library list](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries.ts).
4. `tags` (OPTIONAL, list[string]). Tags with keywords to describe the project. There can be multiple tags.
5. `owners` (list). There can be multiple owners. For each owner the following fields are present.

    1. `organization` (REQUIRED, string). Name of the organization that owns the model.
    2. `name` (OPTIONAL, string). Name of a contact person within the organisation.
    3. `email` (OPTIONAL, string). Email address of the contact person or organization.
    4. `role` (OPTIONAL, string). Role of the contact person. This field should only be set when the `name` field
    is set.

### 2. `model_index`

There can be multiple models. For each model the following fields are present.

> 1. `name` (REQUIRED, string). The name of the model.

Author: A URI pointing to a model file would be a good addition. For the other points: I don't know. I really depends what kind of information we want to capture. So the answer can be yes as well as no.

> 1. `name` (REQUIRED, string). The name of the model.

Author: Added a field that contains an URI to a model file. Did not add any other fields, this is maybe premature.

2. `results` (list). There can be multiple results. For each result the following fields are present.

    1. `task` (list).

        1. `task_type` (REQUIRED, string). The task of the model, for example "object-classifcation".
        2. `task_name` (OPTIONAL, string). A pretty name fo the model taks, for example "Object Classification".

>     2. `Dataset` (list).

Reviewer: ```suggestion
    2. `datasets` (list).
```

         - class: {class_name}                          # Optional. Name of the output class the graph depends on.
           feature: {feature_name}                      # Required. Name of the feature the graph depends on.
           data:
            - x_value: {x_value}                        # Required. The x value of the graph data.
              y_value: {y_value}                        # Required. The y value of the graph data.

# Assessments like IAMA.
assessments:
- name: {assessment_name}                               # Required. Example: IAMA.
  date: {assessment_date}                               # Required. Example: 25-03-2025.
  contents:
    - question: {question_text}                         # Required. Example: "Question 1: ...".
      answer: {answer_text}                             # Required. Example: "Answer: ...".
      remarks: {remarks_text}                           # Optional. Example: "Remarks: ...".
      authors:                                          # Optional. Example: "['John', 'Peter']".
        - name: {author_name}
      timestamp: {timestamp}                            # Optional. Example: 1711630721.
```

## Schema

```json

> {

Reviewer 4: could you add an $id

>     "$schema": "http://json-schema.org/draft-06/schema#",

Reviewer 4: draft-07 or 2020-12 for json schema

    "$ref": "#/definitions/TadSchema",
    "definitions": {
        "TadSchema": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "language": {
                    "type": "array",
                    "items": {
                        "type": "string"
                    }
                },
                "licence": {

>                     "type": "string"

Reviewer 4: Should this be an enum

Face metadata specification.
2. Capturing additional *measurements* on fairness and bias, which can be partitioned into bar plot like
measurements (such as mean absolute SHAP values) and graph plot like measurements (such as partial dependence). This is achieved
by defining an new field `measurements`.
3. Capturing *assessments* (such as IAMA). This is achieved by defining a new field `assessments`.

This standard does not contain all fields present in the Hugging Face metadata specification. The fields that
are optional in the HuggingFace specification AND are specific to the HugginFace interface are ommited.

Following Hugging Face, this proposed standard will be written in yaml.

## Intended usage

The reporting standard defined in this document is intented to be given as a preamble in a model card, similar to
the Hugging Face metadata.

## Specification of the standard

The standard will be written in yaml. An example yaml is given in the next section. The standard contains the
following information.

1. `algorithm-registers` (OPTIONAL, list[string]). If this algorithm is registered in an algorithm register,

> this field should contain a URI to the corresponding record in the register. There can be multiple registers.

Reviewer: ```suggestion
this field should contain a URI to the corresponding record in the register. There can be multiple registers. A URI could look like `https://huggingface.co/org/modelid`.
```

Author: Is there a reason your example URI is huggingface and not from algoritmeregister?

Reviewer: Maybe include both?

2. `language` (OPTIONAL, list[string]). If relevant, the natural languages the model supports in ISO 639.
    There can be multiple languages.
3. `license`(REQUIRED, string). Any license from the [license list](https://huggingface.co/docs/hub/repositories-licenses).
If the license is NOT present in the license list, the following two fields will be REQUIRED.

    1. `license_name` (string). An id for the license.
    2. `license_link` (string). A link to a file of that name inside the repo, or a URL to a remote file containing the license
    contents.

4. `library` (OPTIONAL, string). Any library from the [library list](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries.ts).

> 5. `tags` (OPTIONAL, list[string]). Tags with keywords to describe the project. There can be multiple tags.

Reviewer: ```suggestion
5. `tags` (OPTIONAL, list[string]). Tags with keywords to describe the project. There can be multiple tags.
6. `upl` (OPTIONAL, string). If this algorithm is part of a product offered by the Dutch Government, it should contain a URI from the Uniform Product List (https://standaarden.overheid.nl/owms/oquery/UPL-actueel.plain).

```

# TAD Reporting Standard

This document describes the Transparancy of Algorithmic Decision Making (TAD) reporting standard.

## Introduction

> The standard almost [^1] [^2] extends the [Hugging Face model card metadata specification](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1)

Reviewer: (or make a tiny additional PR after this one to refactor and use that extension?)

> The standard almost [^1] [^2] extends the [Hugging Face model card metadata specification](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1)

Reviewer 3: I would include it in this PR because now it's the first time we use it. So I think it makes sense that add it as a single commit in this PR.

to allow for:

> 1. More finegrained information on performance metrics, by extending the `metrics_field` from the Hugging

Reviewer 2: I love this management summary :)

Face metadata specification.
2. Capturing additional *measurements* on fairness and bias, which can be partitioned into bar plot like
measurements (such as mean absolute SHAP values) and graph plot like measurements (such as partial dependence). This is achieved
by defining an new field `measurements`.
3. Capturing *assessments* (such as IAMA). This is achieved by defining a new field `assessments`.

This standard does not contain all fields present in the Hugging Face metadata specification. The fields that
are optional in the HuggingFace specification AND are specific to the HugginFace interface are ommited.

Following Hugging Face, this proposed standard will be written in yaml.

> ## Intended usage

Author: Restructured to System Card + Model Card

The reporting standard defined in this document is intented to be given as a preamble in a model card, similar to
the Hugging Face metadata.

## Specification of the standard

The standard will be written in yaml. An example yaml is given in the next section. The standard contains the
following information.

1. `algorithm-registers` (OPTIONAL, list[string]). If this algorithm is registered in an algorithm register,
this field should contain a URI to the corresponding record in the register. There can be multiple registers.
A URI could look like `https://huggingface.co/org/modelid` or `https://algoritmes.overheid.nl/nl/algoritme/modelid`.
2. `language` (OPTIONAL, list[string]). If relevant, the natural languages the model supports in ISO 639.
    There can be multiple languages.

> 3. `license`(REQUIRED, string). Any license from the [license list](https://huggingface.co/docs/hub/repositories-licenses).

Author: Now that I read your comment, I see that my description is not clear enough. The list of licences has an item "other". If "other" is selected the fields below "licence_name" and "licence_link" will be required, so you can refer to a license file in the project repo for example. I will make those fields optional as well, so we can have projects without a license.

> 3. `license`(REQUIRED, string). Any license from the [license list](https://huggingface.co/docs/hub/repositories-licenses).

Author: Actually initially I referred to this list. The thing is, if we wan't to be compatible with HF, we need to use their list.

    contents.

4. `library` (OPTIONAL, string). Any library from the [library list](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries.ts).
5. `tags` (OPTIONAL, list[string]). Tags with keywords to describe the project. There can be multiple tags.
6. `upl` (OPTIONAL, string). If this algorithm is part of a product offered by the Dutch Government,
    it should contain a URI from the [Uniform Product List](https://standaarden.overheid.nl/owms/oquery/UPL-actueel.plain).
7. `owners` (list). There can be multiple owners. For each owner the following fields are present.

    1. `organization` (REQUIRED, string). Name of the organization that owns the model.
    2. `oin` (OPTIONAL, string). If applicable the [Organisatie-identificatienummer (OIN)](https://www.logius.nl/domeinen/toegang/organisatie-identificatienummer/wat-is-het).
    3. `name` (OPTIONAL, string). Name of a contact person within the organisation.
    4. `email` (OPTIONAL, string). Email address of the contact person or organization.
    5. `role` (OPTIONAL, string). Role of the contact person. This field should only be set when the `name` field
    is set.

> ### 2. `model_index`

Author: The reason I named this `model_index` is to follow the specification of Hugging Face. I agree the name "models" makes more sense.

> ### 2. `model_index`

Reviewer 3: @anneschuth in the TAD reporting standard we try not to report about a single model. What we try to achieve is to report about the project / AI system. An example is that we added the section `assessments` and the suggestion I made to add datasets as well.
That's why I don't agree with sticking to `model_index` but propose to use `models`.

Another reason why I think we should switch to `models` is that in a project we can have multiple models where we want to report about. What could be an idea is that in the project/system card we have a list of models. These models can refer to individual repos where a `model_index` is available.

> ### 2. `model_index`

Author: We are backwards compatible with hugging face, with exception with the things I put in the footnotes. I don't really mind calling it "models" versus "model-index", with the caveat that we then are deviating from the HF specification (which is fine if we don't care too much about that).

> ### 2. `model_index`

Reviewer: `model-index` already allows for multiple models

There can be multiple models. For each model the following fields are present.

1. `name` (REQUIRED, string). The name of the model.
2. `model` (REQUIRED, string). A URI pointing to a repository containing the model file.
3. `results` (list). There can be multiple results. For each result the following fields are present.

    1. `task` (list).

        1. `task_type` (REQUIRED, string). The task of the model, for example "object-classifcation".
        2. `task_name` (OPTIONAL, string). A pretty name fo the model taks, for example "Object Classification".

    2. `dataset` (list).

        1. `type` (REQUIRED, string). The type of the dataset, can be a dataset id from [HuggingFace datasets](https://hf.co/datasets)
        or a link to a repository containing the dataset[^1], for example "common_voice".

>         2. `name` (REQUIRED, string). Name pretty name for the dataset, for example "Common Voice (French)".

Reviewer 2: Just like with the metrics i could also see that we maybe want to show some features of the dataset, but this is more for a dataset modelcard to be honest

Author: We could include it. Anne also did it in in the modelcard in his POC.

>         3. `split` (OPTIONAL, string). The split of the dataset, for example "train".

Reviewer 2: I would also like to have a link to the dataset if it is available (like huggingface actually also does like for this model https://huggingface.co/NetherlandsForensicInstitute/vuurwerkverkenner)

Reviewer 2: I see also that at the bottom you say we deviate from the huggingface deviation but i don't see the id back in here though?

Author: Actually `type` contains a link to the dataset, for hugging faces these are in https://hf.co/datasets. The deviation comes from the fact that we want to allow for other datasets than those on hugging faces. So either we REQUIRE this field and allow it to contain URI's pointing to a dataset OR we make it OPTIONAL and make a new field with a URI to a dataset. In either case we deviate from hugging face, because in the first case we allow other things than HF dataset IDs and in the other case we change the field from REQUIRED to OPTIONAL.

        4. `revision` (OPTIONAL, string). Version of the dataset, for example 5503434ddd753f426f4b38109466949a1217c2bb.

    3. `metrics` (list). There can be multiple metrics. For each metric the following fields are present.

        1. `type` (REQUIRED, string). A metric-id from [Hugging Face metrics](https://hf.co//metrics)[^2], for example accuracy.
        2. `name` (REQUIRED, string). A descriptive name of the metric. For example "false positive rate" is
        not a descriptive name, but "training false positive rate w.r.t class x" is.
        3. `value` (REQUIRED, float). The value of the metric.
        4. `class` (OPTIONAL, string/int/float/bool). Some metrics (such as false positive rate for multiclass classification)
        depend on a specific output class. In this field the output class can be specified. It must only be set for metrics
        for which it makes sense.

>         5. `subgroups` (list). Metrics can be computed on subgroups of specific features. For example one can compute the

Author: Hmm, can you give an example of a metric that would only be relevant for certain model parameters?

>         5. `subgroups` (list). Metrics can be computed on subgroups of specific features. For example one can compute the

Reviewer 3: @ChristopherSpelt an example can be that you want to store metrics per epoch to report about development of a metric during the training of a model. This could be used as part of validating a model.

        accuracy for examples where the feature "gender" is set to "male".
        There can be multiple subgroups, which means that the metric is computed on the intersection of those subgroups.
        For each subgroup the following fields are present.

            1. `feature` (OPTIONAL, string). The name of the feature. For example: "gender".
            2. `group` (OPTIONAL, string). The value of the feature. If `feature` is set, this field must be set as wel.
            For example: "male".

>     4. `measurements`.

Reviewer 2: ```suggestion
    4. `measurements for plots`.
```

Reviewer 2: I don't really agree with my suggestion, but I would like to make measurements a bit more specific that it is purely for plots

Reviewer: I don't think it should necessarily be just for plots (also, I don't think we should/can have keys with spaces)

>         1. `bar_plots` (list). The purpose of this field is to capute bar plot like measurements, for example SHAP values.

Reviewer 2: I think these sections are pretty interesting :) can't wait to see it more into action, as i think for sure measurements are way better then putting in binary/images and in this way we could also more easily compare between different models

Author: Yes, I am also very curious to see if this actually captures information at the right granularity!

license_link: {license_link}                            # Required. if license = other, specify "LICENSE" or "LICENSE.md" to link to a file of that name inside the repo, or a URL to a remote file.
library_name: {library_name}                            # Optional. Example: keras or any library from https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/model-libraries.ts.
tags:
- {tag_0}                                               # Optional. Example: audio
- {tag_1}                                               # Optional. Example: automatic-speech-recognition
upl: {upl_uri}                                          # Optional. Example: https://standaarden.overheid.nl/owms/terms/AangifteVertrekBuitenland
owners:
- organization: {organization_name}                     # Required. Example: BZK
  oin: {oin}                                            # Optional. Example: 00000001003214345000
  name: {owner_name}                                    # Optional. Example: John Doe
  email: {owner_email}                                  # Optional. Example: johndoe@email.com
  role: {owner_role}                                    # Optional. Example: Data Scientist.

> model-index:

Reviewer 3: What do we think of adding a section like "parameters"? Similar to metrics, but the parameters contain information about the configuration of the trained model.

Reviewer: I would very much like this

> model-index:

Reviewer 3: Similar question. What do we think of adding a section `artifacts` where we can refer with a URI to a repository containing relevant artifact files for a model?

Reviewer: And this too

- name: {model_id}                                      # Required. Example: CatClassifier.
  model: {model_uri}                                    # Required. URI to a repository containing the model file.
  results:
  - task:
      type: {task_type}                                 # Required. Example: image-classification.
      name: {task_name}                                 # Optional. Example: Image Classification.
    dataset:
      type: {dataset_type}                              # Required. Example: common_voice. Link to a repository containing the dataset or a dataset id from https://hf.co/datasets.
      name: {dataset_name}                              # Required. Example: "Common Voice (French)". A pretty name for the dataset.
      revision: {dataset_version}                       # Optional. Example: 5503434ddd753f426f4b38109466949a1217c2bb
    metrics:

>     - type: {metric_type}                               # Required. Example: false-positive-rate. Use metric id from https://hf.co/metrics.

Author: First of all the terminology is a bit unfortunate, but that's because I stick to the terminology used by hugging face. They use "type" to denote their metric type, which is not a data type. I can add a field "dtype" to specify the datatype and add a suggestion that "name" can be the same as "type" which is the metric id.

      name: {metric_name}                               # Required. Example: "FPR wrt class 0 restricted to feature gender:0 and age:21".
      value: {metric_value}                             # Required. Example: 0.75.
      class: {metric_output_class}                      # Optional. Example: "1". Only relevant for certain metrics such as for example false-positive-rate positive rate.
      subgroups:
        - feature: {feature_name}                       # Optional. Example: "gender".
          group: {feature_subgroup}                     # Optional. Example: "female".

>     measurements:

Reviewer 3: What is the difference between a measurement and a metric?

Author: Good question. A metric is a performance metric, usually a scalar, things like accuracy, recall, precision, false positive rate, etc. Measurement is a term conjured up by Anne and me to name things like SHAP, Partial Dependence, etc. Those things aren't metrics.

Reviewer: It didn't come out of thin air though: https://www.nist.gov/itl/ssd/software-quality-group/metrics-and-measures#Metrics_vs._Measures

Reviewer 3: @ChristopherSpelt could it be that what you have in mind with measurements are related to a model in a different way? What do you think if you read the rest of my comment?

If I see the examples for SHAP and Partial Dependence, I think these are metrics that are informative about the predicted outcome of a machine learning model. If you run these tools in the training phase of a model, the result is a metric so we could store them as a metric.

When we have a model in production and we have predicted outcomes, we also want to be able to log these predicted outcomes in a project/system card. In that section of the card I think using the term `measurement` might be something we can consider.

Author: I am not sure if I understand what you are suggesting? The reason why we chose to make a different category called "measurements" is because the nature of some of the fairness tests such as shap, pd, etc are not metrics and cannot be logically captured in that field.

Reviewer: Right. That would be yet another thing we can measure. Wouldn't we call something like that `monitoring` or `logging`?

Reviewer 3: We could align on this during our meeting I proposed for the `model_index`. To make it easier to apply the format, I trying to figure out what is the most logical way to define the `measurements`. I try to prevent that we have to create a new minor version of the reporting standard every time we have a new type of measurement we want to include.

>       # Bar plots should be able to capture SHAP and Robustness Toolbox from AI Verify.

Reviewer 3: Is SHAP a logical example? Because SHAP is used to explain the output of a machine learning model. I.s.o. the "model card" I would expect to have such an instrument listed for something like a "output card". What do you think?

Reviewer: I'm not sure I understand this comment. What is an "output card"?

Author: SHAP can be used to explain how important features are on for example train or test data (as far as my understanding goes). My assumption was that we would like to capture these kind of test results in our model card.  At this point I would just stick to one thing, which we can call a "model card", and in the future we will probably split the fields among different "cards".

      bar_plots:
      - type: {measurement_type}                        # Required. Example: "SHAP".

>         name: {measurement_name}                        # Optional. Example: "Mean Absoulte Shap Values".

Reviewer 2: ```suggestion
        name: {measurement_name}                        # Optional. Example: "Mean Absolute Shap Values".
```
